### PR TITLE
Npm as package manager

### PR DESCRIPTION
**Pull request by Fatme**

### CODE REVIEW

File: lib/node-package-manager.ts [hunk 1]
```diff
@@ -50,128 +26,51 @@ export class NodePackageManager implements INodePackageManager {
...
-				}
-
-				if(!this.isPackageUnpacked(packagePath).wait()) {
-					this.cacheUnpack(packageName, version).wait();
-				}
-				return packagePath;
-			}
-		}).future<string>()();
+	public cache(packageName: string, version: string, config?: any): IFuture<ICacheData> {
+		// function cache (pkg, ver, where, scrub, cb)
```
Comment (rosen-vladimirov) [inline lib/node-package-manager.ts]: do we need this comment?

File: lib/node-package-manager.ts [hunk 2]
```diff
@@ -50,128 +26,51 @@ export class NodePackageManager implements INodePackageManager {
...
-		return future;
+	
+	public cacheUnpack(packageName: string, version: string, unpackTarget?: string): IFuture<void> {
+		// function unpack (pkg, ver, unpackTarget, dMode, fMode, uid, gid, cb)
+		return this.loadAndExecute("cache", [packageName, version, unpackTarget, null, null, null, null], { subCommandName: "unpack" });
 	}
-
-	private isPackageCached(packagePath: string): IFuture<boolean> {
-		return this.$fs.exists(packagePath);
+
```
Comment (rosen-vladimirov) [inline lib/node-package-manager.ts]: some space left here

File: lib/node-package-manager.ts [hunk 3]
```diff
@@ -50,128 +26,51 @@ export class NodePackageManager implements INodePackageManager {
...
-
-	private isPackageCached(packagePath: string): IFuture<boolean> {
-		return this.$fs.exists(packagePath);
+	
+	public view(packageName: string, propertyName: string): IFuture<any> {
+		return this.loadAndExecute("view", [[packageName, propertyName], [false]]);
 	}
-
-	private isPackageUnpacked(packagePath: string): IFuture<boolean> {
+
```
Comment (rosen-vladimirov) [inline lib/node-package-manager.ts]: some space left here

File: lib/node-package-manager.ts [hunk 4]
```diff
@@ -50,128 +26,51 @@ export class NodePackageManager implements INodePackageManager {
...
-
-		return future;
+	
+	public install(packageName: string, pathToSave: string, config?: any): IFuture<any> {
+		return this.loadAndExecute("install", [pathToSave, packageName], { config: config });
 	}
-
-	public getCachedPackagePath(packageName: string, version: string): string {
-		return path.join(npm.cache, packageName, version, "package");
+
```
Comment (rosen-vladimirov) [inline lib/node-package-manager.ts]: some space left here

File: lib/node-package-manager.ts [hunk 5]
```diff
@@ -50,128 +26,51 @@ export class NodePackageManager implements INodePackageManager {
...
+	
+	public install(packageName: string, pathToSave: string, config?: any): IFuture<any> {
+		return this.loadAndExecute("install", [pathToSave, packageName], { config: config });
 	}
-
-	public getCachedPackagePath(packageName: string, version: string): string {
-		return path.join(npm.cache, packageName, version, "package");
+	
+	public uninstall(packageName: string, config?: any): IFuture<any> {
+		return this.loadAndExecute("uninstall", [[packageName]], { config: config });
```
Comment (rosen-vladimirov) [inline lib/node-package-manager.ts]: why do you have double [ around the package name, shouldn't it be [packageName]?

File: lib/services/platform-service.ts [hunk 6]
```diff
@@ -169,14 +177,17 @@ export class PlatformService implements IPlatformService {
...
+			// Process node_modules folder
+			this.$pluginsService.ensureAllDependenciesAreInstalled().wait();
+			var tnsModulesDestinationPath = path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME, PlatformService.TNS_MODULES_FOLDER_NAME);
+			this.$broccoliBuilder.prepareNodeModules(tnsModulesDestinationPath, this.$projectData.projectDir).wait();
 
+			this.$logger.out("Project successfully prepared");
 		}).future<void>()();
 	}
-
+
```
Comment (rosen-vladimirov) [inline lib/services/platform-service.ts]: space left

File: lib/services/plugins-service.ts [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+			let nodeModules = this.getDependencies();
+			return _.map(nodeModules, nodeModuleName => this.getNodeModuleData(nodeModuleName).wait());
+		}).future<INodeModuleData[]>()();
+	}
+	
+	private executeNpmCommand(npmCommandName: string, npmCommandArguments: string): IFuture<string> {
+		return (() => {
+			let result = "";
+			
+			if(npmCommandName === PluginsService.INSTALL_COMMAND_NAME) {
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: You can use enum for npmCommandName values, but this is not a show stopper

File: lib/services/plugins-service.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	private parseNpmCommandResult(npmCommandResult: string): string {  // [[name@version, node_modules/name]]
+		return npmCommandResult[0][0].split("@")[0]; // returns plugin name
+	}
+	
+	private executeForAllInstalledPlatforms(action: (pluginDestinationPath: string, platform: string, platformData: IPlatformData) => void): void {
+		let availablePlatforms = _.keys(this.$platformsData.availablePlatforms);
+		_.each(availablePlatforms, platform => {
+			let isPlatformInstalled = this.$fs.exists(path.join(this.$projectData.platformsDir, platform.toLowerCase())).wait();
+			if(isPlatformInstalled) {
+				let platformData = this.$platformsData.getPlatformData(platform.toLowerCase());
```
Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: you can move `platform.toLowerCase()` outside of the each.

Comment (rosen-vladimirov) [inline lib/services/plugins-service.ts]: ahh, no, there's duplication of method's action paramater and each's argument - both are platform, which led me to this confusion. Consider renaming each's platform to availablePlatform or avlblPlatform or smth else.

Comment (Fatme) [inline lib/services/plugins-service.ts]: I've fixed this.

File: lib/services/plugins-service.ts (hunk 7 above)
Comment (Fatme) [inline lib/services/plugins-service.ts]: I think static constants is better.

File: lib/node-package-manager.ts (hunk 2 above)
Comment (Fatme) [inline lib/node-package-manager.ts]: I've tried to fix this but github doesn't show me the file as changed.

**Comment by rosen-vladimirov** [comment]
:+1:  (you can squash some of the commits :smile: )
and great work 
baloo_thumbs_up_with_style_by_rpiquel-d5iajxi

**Pull request merged**